kdc/kerberos5.c

```diff
@@ -799,22 +851,30 @@ pa_enc_ts_validate(astgs_request_t r, const PA_DATA *pa)
 		   "(enctype %s) error %s",
 		   r->cname, str ? str : "unknown enctype", msg);
 	krb5_free_error_message(r->context, msg);
-	free(str);
+
+	if (auth_state->free_ptr != NULL) {
+	    free(auth_state->free_ptr);
+	}
+	auth_state->auth_status = HDB_AUTH_WRONG_PASSWORD;
+	auth_state->auth_details = str ? str : "unknown enctype";
+	auth_state->free_ptr = str;
 
 	if(hdb_next_enctype2key(r->context, &r->client->entry, NULL,
 				enc_data.etype, &pa_key) == 0)
 	    goto try_next_key;
 
 	free_EncryptedData(&enc_data);
 
-	if (r->clientdb->hdb_auth_status)
-	    r->clientdb->hdb_auth_status(r->context, r->clientdb, r->client,
-					 HDB_AUTH_WRONG_PASSWORD);
-
 	ret = KRB5KDC_ERR_PREAUTH_FAILED;
 	goto out;
     }
     free_EncryptedData(&enc_data);
+    if (auth_state->free_ptr != NULL) {
+	    free(auth_state->free_ptr);
+    }
+    *auth_state = (struct kdc_pa_auth_state) {
+	.auth_status = -1,
```

Comment:
Is this portable to all C compilers used to build Heimdal?

Reply:
Any clues on what the more exotic examples would be?  (This passed Samba CI on our platforms, but those are fairly modern)

Reply:
Just checking if C99 is our baseline, if so then designated initialisers are OK.

Reply:
It's not, sadly.  @jaltman still needs to support older Microsoft compilers that don't support designated initializers.  I know, it's awful.  I desperately want to be able to use designated initializers!